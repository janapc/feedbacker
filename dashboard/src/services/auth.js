export default (httpClient) => ({
  register: async ({ name, email, password }) => {
    const response = await httpClient.post('/auth/register', { name, email, password });

    let errors = null;

    if (!response.data) {
      errors = {
        status: response.request.status,
        statusText: response.request.statusText,
      };
    }

    return {
      data: response.data,
      errors,
    };
  },
  login: async ({ email, password }) => {
    const response = await httpClient.post('/auth/login', { email, password });
    let errors = null;

    if (!response.data) {
      errors = {
        status: response.request.status,
        statusText: response.request.statusText,
      };
    }

    return {
      data: response.data,
      errors,
    };
  },
});
