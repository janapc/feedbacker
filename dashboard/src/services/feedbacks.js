const defaultPagination = {
  limit: 5,
  offset: 0,
};

export default (httpClient) => ({
  getAll: async ({ type, limit, offset } = defaultPagination) => {
    const query = { limit, offset };
    if (type) {
      query.type = type;
    }
    const response = await httpClient.get('/feedbacks', { params: query });

    return { data: response.data };
  },
  getSummary: async () => {
    const response = await httpClient.get('/feedbacks/summary');

    return { data: response.data };
  },
});
