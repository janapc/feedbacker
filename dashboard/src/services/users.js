export default (httpClient) => ({
  getMe: async () => {
    const response = await httpClient.get('/users/me');

    return {
      data: response.data,
    };
  },
  generateApiKey: async () => {
    const response = await httpClient.post('/users/me/apikey');

    return {
      data: response.data,
    };
  },
});
