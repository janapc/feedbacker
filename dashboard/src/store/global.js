import { reactive } from 'vue';

const globalInitialState = {
  isLoading: false
};

const state = reactive(globalInitialState);

export default state;

export function resetGlobalStore () {
  state.isLoading = false;
}

export function setGlobalLoading (status) {
  state.isLoading = status;
}
