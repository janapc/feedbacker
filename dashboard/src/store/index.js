import { readonly } from 'vue';
import UserModule from './user';
import GlobalModule from './global';

export default readonly({
  User: UserModule,
  Global: GlobalModule,
});
