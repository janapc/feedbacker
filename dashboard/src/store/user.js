import { reactive } from 'vue';

const userInitialState = {
  currentUser: {}
};

const state = reactive(userInitialState);

export default state;

export function resetUserStore () {
  state.currentUser = {};
}

export function cleanCurrentUser () {
  state.currentUser = {};
}

export function setCurrentUser (user) {
  state.currentUser = user;
}

export function setApiKey (apiKey) {
  const currentUser = { ...state.currentUser, apiKey };
  state.currentUser = currentUser;
}
