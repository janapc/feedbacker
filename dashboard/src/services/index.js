import axios from 'axios';

import router from '../router';
import AuthService from './auth';
import UsersService from './users';
import FeedbacksService from './feedbacks';

import { setGlobalLoading } from '../store/global';

const API_ENVS = {
  production: 'https://backend-vue3js-ixvpypm03.vercel.app/',
  development: 'http://localhost:3000',
  local: 'http://localhost:3000',
};

const httpClient = axios.create({
  baseURL: API_ENVS[process.env.NODE_ENV] ?? API_ENVS.local,
});

httpClient.interceptors.response.use(
  (response) => {
    setGlobalLoading(false);
    return response;
  },
  (error) => {
    const canThrowAnError = error.request.status === 0 || error.request.status === 500;

    if (canThrowAnError) {
      setGlobalLoading(false);
      throw new Error(error.message);
    }
    if (error.response.status === 401) {
      router.push({ name: 'Home' });
    }
    setGlobalLoading(false);
    return error;
  },
);

httpClient.interceptors.request.use((config) => {
  setGlobalLoading(true);
  const token = window.localStorage.getItem('token');

  if (token) {
    config.headers.common.Authorization = `Bearer ${token}`;
  }
  return config;
});

export default {
  auth: AuthService(httpClient),
  users: UsersService(httpClient),
  feedbacks: FeedbacksService(httpClient),
};
