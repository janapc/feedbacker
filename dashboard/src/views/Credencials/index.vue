<template>
  <div class="flex justify-center w-full h-28 bg-brand-main">
    <header-logged id="headerLogged" />
  </div>

  <div class="flex flex-col items-center justify-center h-64 bg-brand-gray">
    <h1 class="text-4xl font-black text-center text-gray-800">Credenciais</h1>
    <p class="text-lg text-center text-gray-800 font-regular">
      Guia de instalação e geração de suas credenciais
    </p>
  </div>

  <div class="flex justify-center w-full h-full">
    <div class="flex flex-col w-4/5 max-w-6xl py-10">
      <h1 class="text-3xl font-black text-brand-darkgray">instalação e configuração</h1>

      <p class="mt-10 text-lg text-gray-800 font-regular">
        Este aqui é a sua chave de api
      </p>

      <content-loader
        id="contentLoader"
        v-if="store.Global.isLoading || state.isLoading"
        class="rounded"
        height="50px"
        width="600px"
      />

      <div
        id="contentApikey"
        v-else
        class="flex py-3 pl-5 mt-2 rounded justify-between items-center bg-brand-gray w-full lg:w-1/2"
      >
        <span v-if="state.hasError" id="error-apiKey">Erro ao carregar a apiKey</span>
        <span v-else id="apikey">{{ String(store.User.currentUser.apiKey) }}</span>

        <div class="flex ml-20 mr-5" v-if="!state.hasError">
          <icon
            @click="handleCopy"
            name="copy"
            id="apikey-copy"
            :color="brandColors.graydark"
            size="24"
            class="cursor-pointer"
          />
          <icon
            id="generate-apikey"
            @click="handleGenerateApiKey"
            name="loading"
            :color="brandColors.graydark"
            size="24"
            class="cursor-pointer ml-3"
          />
        </div>
      </div>

      <p class="mt-5 text-lg text-gray-800 font-regular">
        Coloque o script abaixo no seu site para começar a receber feedbacks
      </p>

      <content-loader
        id="content-loader"
        v-if="store.Global.isLoading || state.isLoading"
        class="rounded"
        height="50px"
        width="600px"
      />

      <div
        id="contentScript"
        v-else
        class="py-3 pl-5 pr-20 mt-2 rounded bg-brand-gray w-full lg:w-2/3 overflow-x-scroll"
      >
        <span v-if="state.hasError" id="error-script">Erro ao carregar o script</span>
        <pre v-else>
&lt;script src="http://janapc-feeadbacker-widget.netlify.app?api_key={{
            String(store.User.currentUser.apiKey)
          }}"&gt;&lt;/script&gt;</pre>
      </div>
    </div>
  </div>
</template>

<script>
import { reactive, watch } from 'vue';

import { useToast } from 'vue-toastification';
import HeaderLogged from '../../components/HeaderLogged/index.vue';
import ContentLoader from '../../components/ContentLoader/index.vue';
import Icon from '../../components/Icon/index.vue';

import palette from '../../../palette';

import services from '../../services';

import useStore from '../../hooks/useStore';
import { setApiKey } from '../../store/user';

export default {
  components: { HeaderLogged, Icon, ContentLoader },
  setup() {
    const store = useStore();
    const toast = useToast();

    const state = reactive({
      isLoading: false,
      hasError: false,
    });

    function handleError(error) {
      state.isLoading = false;
      state.hasError = !!error;
    }

    watch(
      () => store.User.currentUser,
      () => {
        if (!store.Global.isLoading && !store.User.currentUser.apiKey) {
          handleError(true);
        }
      },
    );

    async function handleCopy() {
      toast.clear();
      try {
        await navigator.clipboard.writeText(store.User.currentUser.apiKey);
        toast.success('Copiado!');
      } catch (error) {
        handleError(error);
      }
    }

    async function handleGenerateApiKey() {
      try {
        state.isLoading = true;
        const { data } = await services.users.generateApiKey();

        setApiKey(data.apiKey);
        state.isLoading = false;
      } catch (error) {
        handleError(error);
      }
    }

    return {
      store,
      state,
      brandColors: palette.brand,
      handleGenerateApiKey,
      handleCopy,
    };
  },
};
</script>

<style></style>
