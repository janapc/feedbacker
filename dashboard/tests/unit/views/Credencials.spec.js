import { shallowMount, flushPromises } from '@vue/test-utils';
import vueToastification, { useToast } from 'vue-toastification';

import services from '../../../src/services';
import Credencials from '../../../src/views/Credencials/index.vue';
import { setApiKey } from '../../../src/store/user';

const mockStore = {
  User: { currentUser: { name: 'Banana', apiKey: '123banana' } },
  Global: { isLoading: false }
};
jest.mock('../../../src/hooks/useStore', () => () => mockStore);

jest.mock('vue-toastification');

const toast = {
  clear: jest.fn(),
  success: jest.fn()
};

jest.mock('../../../src/services');
jest.mock('../../../src/store/user');

const mockClipboard = {
  writeText: jest.fn()
};

global.navigator.clipboard = mockClipboard;

describe('<Credencials />', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
  });
  it('should render Credencials correctly', () => {
    const wrapper = shallowMount(Credencials, {
      global: {
        plugins: [vueToastification]
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render the content loader if the variable isLoading be true', () => {
    mockStore.Global = { isLoading: true };

    const wrapper = shallowMount(Credencials, {
      global: {
        plugins: [vueToastification]
      }
    });

    expect(wrapper.findComponent('#headerLogged').exists()).toBe(true);
    expect(wrapper.findComponent('#contentLoader').exists()).toBe(true);

    expect(wrapper.find('#contentApikey').exists()).toBe(false);
    expect(wrapper.find('#contentScript').exists()).toBe(false);
  });

  it('should show the API key of the user and generate a script for the user', () => {
    mockStore.Global = { isLoading: false };

    const sriptText = '<script src="http://janapc-feeadbacker-widget.netlify.app?api_key=123banana"></script>';

    const wrapper = shallowMount(Credencials, {
      global: {
        plugins: [vueToastification]
      }
    });

    expect(wrapper.find('#apikey').text()).toBe('123banana');
    expect(wrapper.find('#apikey-copy').attributes('color')).toContain('#C0BCB0');
    expect(wrapper.find('#generate-apikey').attributes('color')).toContain('#C0BCB0');
    expect(wrapper.find('pre').text()).toBe(sriptText);
  });

  it('should copy the api key of user', async () => {
    useToast.mockReturnValueOnce(toast);
    const wrapper = shallowMount(Credencials);

    expect(wrapper.find('#apikey').text()).toBe('123banana');
    expect(wrapper.find('#apikey-copy').exists()).toBeTruthy();

    await wrapper.find('#apikey-copy').trigger('click');

    await flushPromises();

    expect(toast.clear).toBeCalled();
    expect(toast.success).toBeCalledWith('Copiado!');
  });

  it('should return a message of the error to the user', async () => {
    services.users.generateApiKey.mockReturnValueOnce({ error: { status: 500 } });

    const wrapper = shallowMount(Credencials, {
      global: {
        plugins: [vueToastification]
      }
    });

    expect(wrapper.find('#apikey').text()).toBe('123banana');
    expect(wrapper.find('#generate-apikey').exists()).toBeTruthy();

    await wrapper.find('#generate-apikey').trigger('click');

    await flushPromises();

    expect(wrapper.find('#error-apiKey').exists()).toBe(true);
    expect(wrapper.find('#error-apiKey').text()).toBe('Erro ao carregar a apiKey');
    expect(wrapper.find('#apikey').exists()).toBe(false);

    expect(wrapper.find('#error-script').exists()).toBe(true);
    expect(wrapper.find('#error-script').text()).toBe('Erro ao carregar o script');
    expect(wrapper.find('pre').exists()).toBe(false);
  });

  it('should generate new api key to user', async () => {
    const newApiKey = 'new123banana123';

    services.users.generateApiKey.mockReturnValueOnce({ data: { apiKey: newApiKey } });
    setApiKey.mockImplementationOnce((apiKey) => {
      mockStore.User.currentUser.apiKey = apiKey;
    });

    const wrapper = shallowMount(Credencials, {
      global: {
        plugins: [vueToastification]
      }
    });

    expect(wrapper.find('#apikey').text()).toBe('123banana');
    expect(wrapper.find('#generate-apikey').exists()).toBeTruthy();

    await wrapper.find('#generate-apikey').trigger('click');
    expect(setApiKey).toBeCalledWith(newApiKey);

    await flushPromises();

    expect(mockStore.User.currentUser.apiKey).toBe(newApiKey);
    expect(wrapper.find('#apikey').text()).toBe(newApiKey);
  });
});
