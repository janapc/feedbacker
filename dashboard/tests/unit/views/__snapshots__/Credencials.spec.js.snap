// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Credencials /> should render Credencials correctly 1`] = `
<div class="flex justify-center w-full h-28 bg-brand-main">
  <header-logged-stub id="headerLogged"></header-logged-stub>
</div>
<div class="flex flex-col items-center justify-center h-64 bg-brand-gray">
  <h1 class="text-4xl font-black text-center text-gray-800">Credenciais</h1>
  <p class="text-lg text-center text-gray-800 font-regular"> Guia de instalação e geração de suas credenciais </p>
</div>
<div class="flex justify-center w-full h-full">
  <div class="flex flex-col w-4/5 max-w-6xl py-10">
    <h1 class="text-3xl font-black text-brand-darkgray">instalação e configuração</h1>
    <p class="mt-10 text-lg text-gray-800 font-regular"> Este aqui é a sua chave de api </p>
    <div id="contentApikey" class="flex py-3 pl-5 mt-2 rounded justify-between items-center bg-brand-gray w-full lg:w-1/2"><span id="apikey">123banana</span>
      <div class="flex ml-20 mr-5">
        <icon-stub name="copy" id="apikey-copy" color="#C0BCB0" size="24" class="cursor-pointer"></icon-stub>
        <icon-stub id="generate-apikey" name="loading" color="#C0BCB0" size="24" class="cursor-pointer ml-3"></icon-stub>
      </div>
    </div>
    <p class="mt-5 text-lg text-gray-800 font-regular"> Coloque o script abaixo no seu site para começar a receber feedbacks </p>
    <div id="contentScript" class="py-3 pl-5 pr-20 mt-2 rounded bg-brand-gray w-full lg:w-2/3 overflow-x-scroll"><pre>&lt;script defer async onload="init('123banana')" src="http://janapc-feeadbacker-widget.netlify.app/init.js"&gt;&lt;/script&gt;</pre>
    </div>
  </div>
</div>
`;
